app/src/main/java/pm/gnosis/heimdall/views/StepIndicator.kt

```diff
@@ -0,0 +1,346 @@
+import kotlin.math.min
+
+
+class StepIndicator @JvmOverloads constructor(
+    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
+) : LinearLayout(context, attrs, defStyleAttr) {
+
+    private val backgroundPaint: Paint
+
+    private val inactiveInactiveEdgeColor: Pair<Int, Int>
+    private val inactiveActiveEdgeColor: Pair<Int, Int>
+    private val activeInactiveEdgeColor: Pair<Int, Int>
+    private val activeActiveEdgeColor: Pair<Int, Int>
+
+    private val edgeShaders: SparseArray<LinearGradient> = SparseArray()
+
+    init {
+        orientation = HORIZONTAL
+        gravity = Gravity.CENTER
+        backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
+        backgroundPaint.style = Paint.Style.FILL_AND_STROKE
+        backgroundPaint.strokeWidth = resources.getDimension(R.dimen.step_indicator_edge_width)
+
+        inactiveInactiveEdgeColor =
+            ContextCompat.getColor(context, R.color.disabled_button) to ContextCompat.getColor(context, R.color.disabled_button)
+        inactiveActiveEdgeColor =
+            ContextCompat.getColor(context, R.color.disabled_button) to ContextCompat.getColor(context, R.color.safe_green)
+        activeInactiveEdgeColor =
+            ContextCompat.getColor(context, R.color.safe_green) to ContextCompat.getColor(context, R.color.disabled_button)
+        activeActiveEdgeColor =
+            ContextCompat.getColor(context, R.color.safe_green) to ContextCompat.getColor(context, R.color.safe_green)
+    }
+
+    fun updateStep(index: Int, stepState: StepState) {
+        val step = getChildAt(index) as Step
+        step.state = stepState
+        edgeShaders.clear()
+        invalidate()
+    }
+
+    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
+        if (changed) {
+            var childsMargin = if (childCount == 0) 0 else (width - paddingLeft * 2 - getChildAt(0).measuredWidth * childCount) / childCount
+            var left = l + paddingLeft + childsMargin / 2
+            for (i in 0 until childCount) {
+                val child = getChildAt(i)
+
+                val childWidth = child.measuredWidth
+                val childHeight = child.measuredHeight
+
+                child.layout(left + childsMargin / 2, paddingTop, left + childWidth + childsMargin / 2, childHeight + paddingTop)
+                left += childWidth + childsMargin / 2
+            }
+        }
+    }
+
+    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
+        super.onSizeChanged(w, h, oldw, oldh)
+        edgeShaders.clear()
+    }
+
+    override fun dispatchDraw(canvas: Canvas) {
+        super.dispatchDraw(canvas)
+
+        var colorPair: Pair<Int, Int>
+
+        for (i in 0 until childCount - 1) {
+
+            val child1 = getChildAt(i) as Step
+            val child2 = getChildAt(i + 1) as Step
+
+            colorPair =
+                if (child1.state == StepState.UNCOMPLETED_INACTIVE || child1.state == StepState.SKIPPED) {
```

Comment:
This can be optimized, but it is fine for now.

E.g. the color of a child doesn't depend on any other child, so you only need 2 colors (active and inactive) and use these in the edge shader

Reply:
good Point